### PR TITLE
Spike chet 161

### PR DESCRIPTION
**Pull request by andrewpowell**
Adding in the approach to use a spring boot app that is configured by the CFN stack to read files from the SQS queue and write them to a location on the host machine (preferably an EFS mount)

### CODE REVIEW

File: filesystem-processor/pom.xml [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+        <dependency>
+            <groupId>org.springframework.cloud</groupId>
+            <artifactId>spring-cloud-starter-aws-messaging</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.springframework.cloud</groupId>
+            <artifactId>spring-cloud-starter-aws-parameter-store-config</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>com.amazonaws</groupId>
```
Comment (bpar476) [inline filesystem-processor/pom.xml]: Any particular reason to use the AWS SDK version 1?

File: filesystem-processor/src/main/java/com/atlassian/migration/datacenter/fs/processor/services/SQSMessageProcessor.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    private final ObjectMapper objectMapper = new ObjectMapper();
+
+    public SQSMessageProcessor(AmazonS3 s3Client, ThreadPoolTaskExecutor threadPoolTaskExecutor, @Value("${app.jira.file.path}") String jiraHome) {
+        this.s3Client = s3Client;
+        this.threadPoolTaskExecutor = threadPoolTaskExecutor;
+        this.jiraHome = jiraHome;
+    }
+
+
+    @SqsListener(QUEUE_LOGICAL_NAME)
```
Comment (bpar476) [inline filesystem-processor/src/main/java/com/atlassian/migration/datacenter/fs/processor/services/SQSMessageProcessor.java]: Oh wow that's neat https://hello.atlassian.net/wiki/spaces/WPT/blog/2020/03/07/658091315/Workplace+Productivity+COVID-19+and+Working+From+Home#Computer-Accessories%2FInternet-Access-Expenses

Comment (bpar476) [inline filesystem-processor/src/main/java/com/atlassian/migration/datacenter/fs/processor/services/SQSMessageProcessor.java]: Oh wow that's neat https://cloud.spring.io/spring-cloud-static/spring-cloud-aws/1.2.3.RELEASE/multi/multi__messaging.html#_annotation_driven_listener_endpoints

File: filesystem-processor/src/main/java/com/atlassian/migration/datacenter/fs/processor/services/SQSMessageProcessor.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        log.debug("Received SQS message {}", s3EventNotificationRecord.toJson());
+        List<S3EventNotificationRecord> s3EventNotificationRecords = s3EventNotificationRecord.getRecords();
+        log.debug("Received ".concat(Integer.toString(s3EventNotificationRecords.size())).concat(" records from S3."));
+        File jiraHomePath = new File(this.jiraHome);
+        if (!jiraHomePath.exists()) {
+            if (jiraHomePath.mkdir()) {
+                log.debug("Created Jira Home path ".concat(jiraHomePath.getAbsolutePath()));
+            }
+        }
+        if (s3EventNotificationRecords.size() == 1) {
```
Comment (bpar476) [inline filesystem-processor/src/main/java/com/atlassian/migration/datacenter/fs/processor/services/SQSMessageProcessor.java]: What's the reason for not just using `forEach` in all cases?

File: migration-helper.yml [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+        Statement:
+          - Effect: Allow
+            Principal:
+              AWS: '*'
+            Action:
+              - SQS:SendMessage
+            Resource: '*'
+            Condition:
+              ArnEquals:
+                aws:SourceArn: "arn:aws:s3:::*"
```
Comment (bpar476) [inline migration-helper.yml]: Does this mean only S3 resources can put events into the queue?

File: src/test/java/com/atlassian/migration/datacenter/core/aws/infrastructure/LocalStackEnvironmentVariables.java [hunk 5]
```diff
@@ -0,0 +1,12 @@
+package com.atlassian.migration.datacenter.core.aws.infrastructure;
+
+import cloud.localstack.docker.annotation.IEnvironmentVariableProvider;
+
+import java.util.Map;
+
+public class LocalStackEnvironmentVariables implements IEnvironmentVariableProvider {
+    @Override
+    public Map<String, String> getEnvironmentVariables() {
```
Comment (bpar476) [inline src/test/java/com/atlassian/migration/datacenter/core/aws/infrastructure/LocalStackEnvironmentVariables.java]: What's the use case for providing these environment variables?

File: migration-helper.yml [hunk 6]
```diff
@@ -0,0 +1,390 @@
+AWSTemplateFormatVersion: 2010-09-09
```
Comment (bpar476) [inline migration-helper.yml]: Is this a duplication of templates/migration-stack/migration-helper.yml?

File: templates/migration-stack/migration-helper.yml [hunk 7]
```diff
@@ -1,48 +1,121 @@
...
+    Properties:
+      Name: /config/migration-helper/app.s3.vpc.endpoints
+      Type: StringList
+      Value: !GetAtt EndpointS3.DnsEntries
+      Description: CloudFormation Stack Name
+
+  StackNameParameter:
+    Type: AWS::SSM::Parameter
+    Properties:
+      Name: /config/migration-helper/cloud.aws.stack.name
```
Comment (bpar476) [inline templates/migration-stack/migration-helper.yml]: I don't see any reference to this parameter in the spring boot app. Has it got an intended use?

File: templates/migration-stack/migration-helper.yml [hunk 8]
```diff
@@ -1,48 +1,121 @@
...
+      Type: String
+      Value: !Ref AWS::StackName
+      Description: CloudFormation Stack Name
+
+  JiraFilePathParameter:
```
Comment (bpar476) [inline templates/migration-stack/migration-helper.yml]: Is the intention that this value can be set by a stack parameter? I think the path will always be `/efs/jira/shared` in the migration stack EC2 instance

File: templates/migration-stack/migration-helper.yml [hunk 9]
```diff
@@ -1,48 +1,121 @@
...
+    Type: AWS::SQS::Queue
+    Properties:
+      DelaySeconds: 0
+      MaximumMessageSize: 262144
+      MessageRetentionPeriod: 864000
+      QueueName: !Sub 'atl-migration-queue-${AWS::StackName}'
+      ReceiveMessageWaitTimeSeconds: 0
+      VisibilityTimeout: 90
+
+  EndpointS3:
```
Comment (bpar476) [inline templates/migration-stack/migration-helper.yml]: Is the VPC endpoint for S3 a performance thing?

File: filesystem-processor/src/main/java/com/atlassian/migration/datacenter/fs/processor/services/SQSMessageProcessor.java (hunk 2 above)
Comment (theghostwhoforks) [reply filesystem-processor/src/main/java/com/atlassian/migration/datacenter/fs/processor/services/SQSMessageProcessor.java]: wrong link @bpar476 ?

Comment (theghostwhoforks) [reply filesystem-processor/src/main/java/com/atlassian/migration/datacenter/fs/processor/services/SQSMessageProcessor.java]: will delete this  comment

**Comment by bpar476** [comment]
Oh and one more question I just thought of:

How do we signal to the message consumer that there are no more files coming?

It's not enough for the queue to be empty as there could be files not uploaded from the source system yet. If we're purely populating the queue via S3 events then we would need to send a "canary" file as the signal.
If we allow other items to be sent on the queue we could send a specific message on the queue that indicates there are no more files being uploaded.

File: filesystem-processor/pom.xml (hunk 1 above)
Comment (andrewpowell) [reply filesystem-processor/pom.xml]: This is what is used by the Spring Cloud AWS project.

File: filesystem-processor/src/main/java/com/atlassian/migration/datacenter/fs/processor/services/SQSMessageProcessor.java (hunk 2 above)
Comment (andrewpowell) [reply filesystem-processor/src/main/java/com/atlassian/migration/datacenter/fs/processor/services/SQSMessageProcessor.java]: 👍

File: migration-helper.yml (hunk 4 above)
Comment (andrewpowell) [reply migration-helper.yml]: correct

File: src/test/java/com/atlassian/migration/datacenter/core/aws/infrastructure/LocalStackEnvironmentVariables.java (hunk 5 above)
Comment (andrewpowell) [reply src/test/java/com/atlassian/migration/datacenter/core/aws/infrastructure/LocalStackEnvironmentVariables.java]: I have a localstack pro license, it gets picked up in the environment variables on the host.

File: migration-helper.yml (hunk 6 above)
Comment (andrewpowell) [reply migration-helper.yml]: Yes

File: templates/migration-stack/migration-helper.yml (hunk 8 above)
Comment (andrewpowell) [reply templates/migration-stack/migration-helper.yml]: Yes, that's the intention. We can use that path as a default.

File: templates/migration-stack/migration-helper.yml (hunk 9 above)
Comment (andrewpowell) [reply templates/migration-stack/migration-helper.yml]: It's a security thing.  Make sure that data only goes over the VPC for S3 calls.  This is not needed now b/c it's included in the ASI

**Comment by andrewpowell** [comment]
So this would need to be installed on the Helper Instance.  It's built as an executable JAR.  I think we could send metadata on the S3 transfer of the "canary file", or last file in the process.

File: migration-helper.yml (hunk 6 above)
Comment (bpar476) [reply migration-helper.yml]: Do we need to keep it?

**Comment by andrewpowell** [comment]
Closing Draft

**Pull request closed without merging**